crypto/s2n_hash.c

```diff
@@ -169,13 +176,301 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+    EVP_MD_CTX *md5_secondary_ctx = state->digest.high_level.evp_md5_secondary.ctx;
+
+    int r;
+    switch (alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_md5(), NULL);
+        break;
+    case S2N_HASH_SHA1:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha1(), NULL);
+        break;
+    case S2N_HASH_SHA224:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha224(), NULL);
+        break;
+    case S2N_HASH_SHA256:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha256(), NULL);
+        break;
+    case S2N_HASH_SHA384:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha384(), NULL);
+        break;
+    case S2N_HASH_SHA512:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha512(), NULL);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestInit_ex(primary_ctx, EVP_sha1(), NULL);
+        r &= EVP_DigestInit_ex(md5_secondary_ctx, EVP_md5(), NULL);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_INIT_FAILED);
+    }
+
+    state->alg = alg;
+
+    return 0;
+}
+
+static int s2n_evp_hash_update(struct s2n_hash_state *state, const void *data, uint32_t size)
+{
+    EVP_MD_CTX *primary_ctx = state->digest.high_level.evp.ctx;
+    EVP_MD_CTX *md5_secondary_ctx = state->digest.high_level.evp_md5_secondary.ctx;
+
+    int r;
+    switch (state->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
+        r = EVP_DigestUpdate(primary_ctx, data, size);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestUpdate(primary_ctx, data, size);
+        r &= EVP_DigestUpdate(md5_secondary_ctx, data, size);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_UPDATE_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_evp_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+{
+    EVP_MD_CTX *primary_ctx = state->digest.high_level.evp.ctx;
+    EVP_MD_CTX *md5_secondary_ctx = state->digest.high_level.evp_md5_secondary.ctx;
+
+    int r;
+    unsigned int digest_size = size;
+    uint8_t expected_digest_size;
+    GUARD(s2n_hash_digest_size(state->alg, &expected_digest_size));
+    eq_check(digest_size, expected_digest_size);
+
+    switch (state->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
+        r = EVP_DigestFinal_ex(primary_ctx, out, &digest_size);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestFinal_ex(primary_ctx, ((uint8_t *) out) + MD5_DIGEST_LENGTH, &digest_size);
```

Comment:
Great find! So, this is only going to affect the `s2n_evp_hash_digest` function (which calls OpenSSL's `EVP_DigestFinal_ex` API). In our other uses, size is either the size of the input data (for `s2n_hash_update`) or is not used within the OpenSSL API's (such as `s2n_low_level_hash_digest`).

I have modified this such that both calls to `EVP_DigestFinal_ex` for `S2N_HASH_MD5_SHA1` use the exact MD5/SHA1 digest size.